### PR TITLE
Auto-Scheduler for Soufflé

### PR DESCRIPTION
**Pull request by SamArch27**
This PR is to merge our recent auto-scheduler for Soufflé.

Currently, users manually find good join orders for rules (also called schedules) and annotate the rules with plan statements. Our auto-scheduler aims to derive good schedules for the user automatically.

Usage:

`souffle <program> -p auto_profile --index-stats`
`souffle -c <program> --auto-schedule=auto_profile`

1. Given a representative input for the first run, it collects selectivity statistics via indexes and writes them to a profile.
2. Soufflé reads the statistics from this profile and passes them to an embedded query optimizer, which finds good schedules using a cost model.

Fundamental Limitations:
1. It only works for finding single-threaded schedules (no consideration of parallelism is involved)
2. It doesn't work well for rules which exit early and produce 0 tuples
3. Subtle interactions between indexing and scheduling can lead to sub-optimal performance (work on buffering and sorting is being done to help alleviate this problem somewhat)
4. It doesn't perform well for very complex recursive workloads

Engineering Limitations:
1. We only collect stats in interpreter mode for now
2. We haven't considered the interactions with provenance yet

### CODE REVIEW

**Comment by b-scholz** [comment]
We need a test-case for auto-scheduling. This should be a functional test-case rather than a performance test (this is hard to implement with our current infrastructure).

File: src/ram/CountUniqueKeys.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+#include <utility>
+#include <vector>
+
+namespace souffle::ram {
+
+/**
+ * @class CountUniqueKeys
+ * @brief Count the number of unique keys on a relation given the key columns etc.
+ *
+ * TODO: Example
```
Comment (b-scholz) [inline src/ram/CountUniqueKeys.h:37]: Please do so.

File: src/ram/CountUniqueKeys.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    CountUniqueKeys* cloning() const override {
+        return new CountUniqueKeys(relation, arity, keyColumns, constantsMap, recursiveRelation);
+    }
+
+protected:
+    void print(std::ostream& os, int tabpos) const override {
+        os << times(" ", tabpos) << "COUNT UNIQUE KEYS " << relation
+           << " RECURSIVE: " << (recursiveRelation ? "true" : "false") << " COLUMNS: " << keyColumns
```
Comment (b-scholz) [inline src/ram/CountUniqueKeys.h:76]: Could you improve the readability of the statement? Normally we don't use `:`. Perhaps you could do two different count statements keys 
```
[REC]UNIQUEKEYCOUNT <relation> A1 [=<constant>], ... 
```
We may want to implement a RAM parser at a later point in time.

File: src/ram/Expression.h [hunk 3]
```diff
@@ -27,6 +27,7 @@ namespace souffle::ram {
 class Expression : public Node {
 public:
     Expression* cloning() const override = 0;
+    virtual ~Expression() = default;
```
Comment (b-scholz) [inline src/ram/Expression.h:30]: Why do we need destructors here? This seems unrelated?

File: src/ram/FloatConstant.h [hunk 4]
```diff
@@ -35,6 +35,8 @@ class FloatConstant : public NumericConstant {
 public:
     explicit FloatConstant(RamFloat val) : NumericConstant(ramBitCast(val)) {}
 
+    virtual ~FloatConstant() = default;
```
Comment (b-scholz) [inline src/ram/FloatConstant.h:38]: Same here?

File: src/ram/NumericConstant.h [hunk 5]
```diff
@@ -38,6 +38,8 @@ class NumericConstant : public Expression {
 protected:
     explicit NumericConstant(RamDomain constant) : constant(constant) {}
 
+    virtual ~NumericConstant() = default;
```
Comment (b-scholz) [inline src/ram/NumericConstant.h:41]: Same here?

**Review by b-scholz — CHANGES_REQUESTED** [review]
- Could you implement the count statements in the synthesiser as well? 
- Could you consider using RAM expressions for constants in the counting statements? This would permit packing/unpacking of constant terms and complex constant arithmetic expressions.

File: src/ram/CountUniqueKeys.h [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    bool equal(const Node& node) const override {
+        const auto& other = asAssert<CountUniqueKeys>(node);
+        return RelationStatement::equal(other) && arity == other.getArity() &&
+               keyColumns == other.getKeyColumns() && constantsMap == other.getConstantsMap() &&
+               recursiveRelation == other.isRecursiveRelation();
+    }
+
+    std::size_t arity;
```
[the remark this answers is not in the document]
Comment (SamArch27) [reply src/ram/CountUniqueKeys.h:91]: We don't. I can get rid of this.

File: src/ram/Expression.h (hunk 3 above)
Comment (SamArch27) [reply src/ram/Expression.h:30]: The COUNT nodes hold a ram::Expression* which under the hood can be a ram::SignedConstant or ram::StringConstant or ...

Since ram::Expression is a polymorphic type, it needs a virtual destructor, right?

Comment (b-scholz) [reply src/ram/Expression.h:30]: I don't believe so since the ram::Node has a virtual destructor. It should be alright.

File: src/ast/analysis/ProfileUse.cpp [hunk 7]
```diff
@@ -32,7 +32,12 @@ namespace souffle::ast::analysis {
 void ProfileUseAnalysis::run(const TranslationUnit&) {
     if (Global::config().has("profile-use")) {
         std::string filename = Global::config().get("profile-use");
-        profile::Reader(filename, programRun).processFile();
+        reader = mk<profile::Reader>(filename, programRun);
+        reader->processFile();
+    } else if (Global::config().has("auto-schedule")) {
+        std::string filename = Global::config().get("auto-schedule");
+        reader = mk<profile::Reader>(filename, programRun);
+        reader->processFile();
```
Comment (b-scholz) [inline src/ast/analysis/ProfileUse.cpp:40]: Push these two lines after the if-then-else

File: src/ast/analysis/ProfileUse.cpp [hunk 8]
```diff
@@ -32,7 +32,12 @@ namespace souffle::ast::analysis {
 void ProfileUseAnalysis::run(const TranslationUnit&) {
     if (Global::config().has("profile-use")) {
         std::string filename = Global::config().get("profile-use");
-        profile::Reader(filename, programRun).processFile();
+        reader = mk<profile::Reader>(filename, programRun);
+        reader->processFile();
```
Comment (b-scholz) [inline src/ast/analysis/ProfileUse.cpp:36]: Delete these two lines

File: src/ast/analysis/ProfileUse.cpp [hunk 9]
```diff
@@ -32,7 +32,12 @@ namespace souffle::ast::analysis {
 void ProfileUseAnalysis::run(const TranslationUnit&) {
     if (Global::config().has("profile-use")) {
```
Comment (b-scholz) [inline src/ast/analysis/ProfileUse.cpp:33]: declare std::string filename before the if-then-else

File: src/ast/analysis/UniqueKeys.cpp [hunk 10]
```diff
@@ -0,0 +1,575 @@
...
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2022, The Souffle Developers. All rights reserved
+ * Licensed under the Universal Permissive License v 1.0 as shown at:
+ * - https://opensource.org/licenses/UPL
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+/************************************************************************
+ *
+ * @file UniqueKeys.cpp
```
Comment (b-scholz) [inline src/ast/analysis/UniqueKeys.cpp:11]: Perhaps you rename the file to ProfilingCostCover (or something similar)?

**Review by b-scholz — CHANGES_REQUESTED** [review]

File: src/synthesiser/Synthesiser.cpp [hunk 11]
```diff
@@ -837,7 +837,7 @@ void Synthesiser::emitCode(std::ostream& out, const Statement& stmt) {
                 if (const auto* signedConstant = as<ram::SignedConstant>(constant)) {
                     value = ramBitCast<RamDomain>(signedConstant->getValue());
                 } else if (const auto* stringConstant = as<ram::StringConstant>(constant)) {
-                    value = ramBitCast<RamDomain>(
+                    value = static_cast<RamDomain>(
```
Comment (OlivierHamel) [inline src/synthesiser/Synthesiser.cpp:840]: This really should be `ramBitCast` (want bit pattern reinterpretation, not numeric coercion), but `convertSymbol2Idx` should return `RamUnsigned` instead of `size_t`.

Comment (SamArch27) [reply src/synthesiser/Synthesiser.cpp:840]: Hi @OlivierHamel, good catch.

I was being lazy and trying to get the compiler to shut up with a static_cast but your suggestion is much better.

Thanks!

Comment (OlivierHamel) [reply src/synthesiser/Synthesiser.cpp:840]: I've a complementary system that works the other way around: Given statistics and functional dependencies for the EDB (currently passed in as pragma metadata), it estimates join sizes using an an abstract interpretation. This is plugged into a an optimizer (best-N literals exhaustive, followed by greedy).

This can be less precise than `--profile` IDB info, but is useful when developing the program since it only requires statistics on the EDB. Ping me if this is something you think would be useful for you folks. It'll require some modifications to bring upstream (this is the main reason why I don't have a PR for it).

Comment (SamArch27) [reply src/synthesiser/Synthesiser.cpp]: Hi @OlivierHamel,

That's really interesting.

I can see the advantage in not having to execute the program (say during rapid prototyping).

Perhaps after this PR is merged you can try out the auto-scheduler in your own system and report back on the performance numbers (both in terms of the auto-scheduling time and the performance of the produced schedules).

**Comment by SamArch27** [comment]
The scheduler test keeps failing.

@b-scholz @XiaowenHu96

File: src/ast/analysis/UniqueKeys.cpp [hunk 12]
```diff
@@ -0,0 +1,575 @@
...
+
+    std::unordered_map<std::size_t, std::set<std::string>> atomIdxToGroundedVars;
+    for (std::size_t i = 0; i < atoms.size(); ++i) {
+        std::set<std::string> groundedVars;
+        visit(*atoms[i], [&](const ast::Variable& v) { groundedVars.insert(v.getName()); });
+        atomIdxToGroundedVars[i] = groundedVars;
+    }
+
+    // #atoms -> variables to join
+    std::map<std::size_t, std::set<std::set<std::size_t>>> cache;
```
Comment (b-scholz) [inline src/ast/analysis/UniqueKeys.cpp:261]: Simplify with user-defined types or better description.

File: src/ast/analysis/UniqueKeys.cpp [hunk 13]
```diff
@@ -0,0 +1,575 @@
...
+
+    std::unordered_map<std::size_t, std::map<std::size_t, const ram::Expression*>> atomToIdxConstants;
```
Comment (b-scholz) [inline src/ast/analysis/UniqueKeys.cpp:263]: Same here.

File: src/ast/analysis/UniqueKeys.cpp [hunk 14]
```diff
@@ -0,0 +1,575 @@
...
+#include <unordered_map>
+#include <unordered_set>
+
+namespace souffle::ast::analysis {
+
+const analysis::PowerSet& UniqueKeysAnalysis::getSubsets(std::size_t N, std::size_t K) const {
+    if (cache.count({N, K})) {
+        return cache.at({N, K});
+    }
+    // result of all combinations
```
[the remark this answers is not in the document]
Comment (SamArch27) [reply src/ast/analysis/UniqueKeys.cpp:52]: This is just a helper function to produce power sets so it's appropriate to keep it using the name PowerSet.

**Review by b-scholz — APPROVED** [review]
Thanks for the contribution!

File: src/synthesiser/Synthesiser.cpp (hunk 11 above)
Comment (OlivierHamel) [reply src/synthesiser/Synthesiser.cpp]: I'll see if I have time; I'd have to backport it since we're stuck on an old fork of Souffle.
